File: src/solution200_299/solution226/Solution.java
package solution200_299.solution226;

/**
 * Script Created by daidai on 2017/7/14.
 */

import structure.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }
        TreeNode tmp = root.left;
        root.left = root.right;
        root.right = tmp;
        invertTree(root.left);
        invertTree(root.right);
        return root;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        TreeNode treeNode = TreeNode.parse(new Integer[]{4, 2, 7, 1, 3, 6, 9});
        solution.invertTree(treeNode);
        TreeNode.levelPrint(treeNode);
    }
}

File: src/solution200_299/solution226/Solution.kt
package solution200_299.solution226

import structure.TreeNode
import java.util.*

/**
 * Script Created by daidai on 2017/7/14.
 */
fun invertTree(treeNode: TreeNode?): TreeNode? {
    if (treeNode == null) {
        return null
    }
    val tmp = treeNode.left
    treeNode.left = treeNode.right
    treeNode.right = tmp;
    invertTree(treeNode.left)
    invertTree(treeNode.right)
    return treeNode
}

fun bfs(treeNode: TreeNode?): TreeNode? {
    if (treeNode == null) {
        return null
    }
    val queue = LinkedList<TreeNode>()
    queue.offer(treeNode)
    while (!queue.isEmpty()) {
        val first = queue.poll()
        val tmp = first.left
        first.left = first.right
        first.right = tmp
        if (first.left != null) {
            queue.offer(first.left)
        }
        if (first.right != null) {
            queue.offer(first.right)
        }
    }
    return treeNode
}

fun main(args: Array<String>) {
    val node = TreeNode.parse(intArrayOf(4, 2, 7, 1, 3, 6, 9).toTypedArray())
    bfs(node)
    TreeNode.levelPrint(node)
}